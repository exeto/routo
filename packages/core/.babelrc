{
  "presets": ["@babel/preset-env"],
  "plugins": [
    ["babel-plugin-import", { "libraryName": "qs" }, "qs"],
    [
      "babel-plugin-import",
      {
        "libraryName": "ramda",
        "libraryDirectory": "src",
        "camel2DashComponentName": false
      },
      "ramda"
    ]
  ]
}
