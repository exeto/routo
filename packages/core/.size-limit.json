[
  {
    "path": "lib/*.js",
    "limit": "10 KB",
    "running": false
  }
]
