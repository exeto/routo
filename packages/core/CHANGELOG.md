# 0.1.2 / 2019-05-09

- Disable encoding for query stringify

# 0.1.1 / 2019-05-02

- Without changes

# 0.1.0 / 2019-04-30

- Initial commit
