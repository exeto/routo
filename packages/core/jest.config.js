'use strict';

module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  collectCoverageFrom: [
    'src/**/*.ts',
    '!src/**/__tests__/**/*.ts',
    '!src/**/*.test.ts',
  ],
};
