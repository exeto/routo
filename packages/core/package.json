{
  "name": "@routo/core",
  "version": "0.2.0",
  "description": "Simple router for Redux",
  "repository": {
    "type": "git",
    "url": "https://github.com/exeto/routo.git",
    "directory": "packages/core"
  },
  "main": "lib/index.js",
  "author": "Timofey Dergachev <t@exeto.me> (https://exeto.me/en)",
  "license": "MIT",
  "scripts": {
    "build": "rm -rf lib && babel src -d lib --ignore **/__tests__,**/*.test.js",
    "codecov": "codecov --token=$CODECOV_TOKEN -F core",
    "size": "yarn build && size-limit",
    "test": "jest --coverage && yarn size",
    "prepublishOnly": "yarn test"
  },
  "files": [
    "lib",
    "src",
    "CHANGELOG.md",
    "LICENSE.md",
    "README.md"
  ],
  "keywords": [
    "history",
    "react",
    "redux",
    "route",
    "router",
    "routing"
  ],
  "dependencies": {
    "history": "^4.7.2",
    "path-to-regexp": "^5.0.0",
    "qs": "^6.7.0",
    "ramda": "^0.26.1"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.4",
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@size-limit/preset-small-lib": "^2.0.2",
    "babel-plugin-import": "^1.11.0",
    "codecov": "^3.3.0",
    "jest": "^24.7.1"
  }
}
