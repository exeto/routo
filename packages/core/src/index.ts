import { createBrowserHistory } from 'history';

import { Route, Router, Options, Listener, LocationDescriptor } from './types';
import { createRouteStorage } from './routes';
import { createState, stringifyQueryParams } from './utils';
import { NOT_FOUND } from './consts';

export { NOT_FOUND, Router };

export const createRouter = (routes: Route[], options?: Options): Router => {
  const notFoundRoute = {
    id: NOT_FOUND,
    path: options?.notFoundPath || '/404',
  };

  const history = options?.history || createBrowserHistory();
  const routeStorage = createRouteStorage(notFoundRoute, routes);
  let state = createState(routeStorage, history.location);
  let listeners: Listener[] = [];

  const notify = () => {
    const currentListeners = listeners.slice();

    currentListeners.forEach((listener) => listener(state));
  };

  history.listen((location, action) => {
    state = {
      ...createState(routeStorage, location),
      action,
      prev: { ...state, prev: null },
    };

    notify();
  });

  const getLocationData = (id: string, data?: LocationDescriptor) => {
    const route = routeStorage.getById(id);
    const pathname = route.createPathname(data?.params || {});
    const search = stringifyQueryParams(data?.queryParams || {});

    return { pathname, search };
  };

  return {
    getState() {
      return state;
    },

    subscribe(fn) {
      listeners.push(fn);

      const unsubscribe = () => {
        const index = listeners.indexOf(fn);

        listeners.splice(index, 1);
      };

      return unsubscribe;
    },

    push(id, data) {
      history.push(getLocationData(id, data));
    },

    replace(id, data) {
      history.replace(getLocationData(id, data));
    },

    markAsNotFound() {
      state = {
        ...state,
        id: NOT_FOUND,
        action: 'REPLACE',
        prev: { ...state, prev: null },
      };

      notify();
    },

    createHref(id, data) {
      const { pathname, search } = getLocationData(id, data);

      return `${pathname}${search}`;
    },

    hasId(id) {
      if (id === NOT_FOUND) {
        return true;
      }

      const route = routeStorage.getById(id);

      return route.id !== NOT_FOUND;
    },
  };
};
