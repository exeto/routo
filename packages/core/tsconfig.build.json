{
  "extends": "../../tsconfig.build.json",
  "include": ["./src/**/*"],
  "exclude": ["**/*.test.*", "**/__tests__/*"],
  "compilerOptions": {
    "outDir": "./lib"
  }
}
