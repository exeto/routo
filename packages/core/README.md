# @routo/core

> Simple framework-agnostic router

## Install

```sh
yarn add @routo/core
```

## Usage

```js
import { createRouter } from '@routo/core';

const HOME = 'router/HOME';
const POST = 'router/POST';

const routes = [
  {
    id: HOME,
    path: '/',
  },
  {
    id: POST,
    path: '/posts/:id',
  },
];

const router = createRouter(routes);

const unsubscribe = router.subscribe(state => console.log(state.pathname));

router.replace(POST, { params: { id: '42' } });
// /posts/42

unsubscribe();
```

## State

```js
{
  id: 'router/POST',
  pathname: '/posts/42',
  search: '',
  queryParams: {},
  action: 'PUSH',
  params: { id: '42' },
  prev: {
    id: 'router/HOME',
    pathname: '/',
    search: '',
    queryParams: {},
    action: null,
    params: {}
    prev: null,
  }
}
```

## License

[MIT](LICENSE.md) Â© [Timofey Dergachev](https://exeto.me)
