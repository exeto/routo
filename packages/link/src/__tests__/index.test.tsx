import '@testing-library/jest-dom';

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { createRouter } from '@routo/core';
import { Provider } from '@routo/react';

import Link from '../index';

describe('Link', () => {
  const HOME = 'router/HOME';
  const POSTS = 'router/POSTS';
  const POST = 'router/POST';

  const routes = [
    {
      id: HOME,
      path: '/',
    },
    {
      id: POSTS,
      path: '/posts',
    },
    {
      id: POST,
      path: '/posts/:id',
    },
  ];

  const createCustomRouter = () => {
    const history = createMemoryHistory();

    return createRouter(routes, { history });
  };

  it('should render simple link', () => {
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={HOME}>Home</Link>
      </Provider>,
    );

    const link = getByText('Home');

    expect(link).toHaveAttribute('href', '/');
  });

  it('should render parameterized link', () => {
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POST} params={{ id: '42' }}>
          Post
        </Link>
      </Provider>,
    );

    const link = getByText('Post');

    expect(link).toHaveAttribute('href', '/posts/42');
  });

  it('should render simple link with query params', () => {
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS} queryParams={{ sort: 'date' }}>
          Posts
        </Link>
      </Provider>,
    );

    const link = getByText('Posts');

    expect(link).toHaveAttribute('href', '/posts?sort=date');
  });

  it('should change router state after click', () => {
    const router = createCustomRouter();

    expect(router.getState()).toMatchSnapshot();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS}>Posts</Link>
      </Provider>,
    );

    fireEvent.click(getByText('Posts'));

    expect(router.getState()).toMatchSnapshot();
  });

  it('should change router state after click with replace action', () => {
    const router = createCustomRouter();

    expect(router.getState()).toMatchSnapshot();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS} action="replace">
          Posts
        </Link>
      </Provider>,
    );

    fireEvent.click(getByText('Posts'));

    expect(router.getState()).toMatchSnapshot();
  });

  it('should render custom component', () => {
    const router = createCustomRouter();

    const Component = (props: any) => <span {...props} data-testid="button" />;

    const { getByTestId } = render(
      <Provider router={router}>
        <Link to={POSTS} component={Component}>
          Posts
        </Link>
      </Provider>,
    );

    const button = getByTestId('button');

    expect(button.tagName).toBe('SPAN');
    expect(button).toHaveAttribute('href', '/posts');
  });

  it('should check className prop', () => {
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS} className="foo">
          Posts
        </Link>
      </Provider>,
    );

    expect(getByText('Posts')).toHaveClass('foo');
  });

  it('should add class when route is active', () => {
    const router = createCustomRouter();

    router.push(POSTS);

    const { getByText, rerender } = render(
      <Provider router={router}>
        <Link to={HOME} activeClassName="foo">
          Home
        </Link>
      </Provider>,
    );

    const link = getByText('Home');

    expect(link).not.toHaveClass('foo');

    rerender(
      <Provider router={router}>
        <Link to={POSTS} activeClassName="foo">
          Posts
        </Link>
      </Provider>,
    );

    expect(link).toHaveClass('foo');
  });

  it('should run onClick prop after click on link', () => {
    const handleClick = jest.fn();
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS} onClick={handleClick}>
          Posts
        </Link>
      </Provider>,
    );

    expect(handleClick).not.toBeCalled();
    fireEvent.click(getByText('Posts'));
    expect(handleClick).toBeCalledTimes(1);
  });

  it('should not handle click when metaKey is pressed', () => {
    const handleClick = jest.fn();
    const router = createCustomRouter();

    const { getByText } = render(
      <Provider router={router}>
        <Link to={POSTS} onClick={handleClick}>
          Posts
        </Link>
      </Provider>,
    );

    const routerState = router.getState();
    const link = getByText('Posts');

    // For prevent errors because jsdom does not support navigation
    link.addEventListener('click', (event) => event.preventDefault());

    fireEvent.click(link, { metaKey: true });
    fireEvent.click(link, { altKey: true });
    fireEvent.click(link, { ctrlKey: true });
    fireEvent.click(link, { shiftKey: true });

    expect(handleClick).not.toBeCalled();
    expect(routerState).toMatchSnapshot();
  });
});
