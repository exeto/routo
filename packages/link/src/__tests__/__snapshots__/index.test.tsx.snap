// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should change router state after click 1`] = `
Object {
  "action": null,
  "id": "router/HOME",
  "params": Object {},
  "pathname": "/",
  "prev": null,
  "queryParams": Object {},
  "search": "",
}
`;

exports[`Link should change router state after click 2`] = `
Object {
  "action": "PUSH",
  "id": "router/POSTS",
  "params": Object {},
  "pathname": "/posts",
  "prev": Object {
    "action": null,
    "id": "router/HOME",
    "params": Object {},
    "pathname": "/",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`Link should change router state after click with replace action 1`] = `
Object {
  "action": null,
  "id": "router/HOME",
  "params": Object {},
  "pathname": "/",
  "prev": null,
  "queryParams": Object {},
  "search": "",
}
`;

exports[`Link should change router state after click with replace action 2`] = `
Object {
  "action": "REPLACE",
  "id": "router/POSTS",
  "params": Object {},
  "pathname": "/posts",
  "prev": Object {
    "action": null,
    "id": "router/HOME",
    "params": Object {},
    "pathname": "/",
    "prev": null,
    "queryParams": Object {},
    "search": "",
  },
  "queryParams": Object {},
  "search": "",
}
`;

exports[`Link should not handle click when metaKey is pressed 1`] = `
Object {
  "action": null,
  "id": "router/HOME",
  "params": Object {},
  "pathname": "/",
  "prev": null,
  "queryParams": Object {},
  "search": "",
}
`;
